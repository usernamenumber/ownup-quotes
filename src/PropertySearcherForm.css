.PropertySearcherForm {
    width: 800px;
    margin-left: 80px;
}
.PropertySearcherForm td {
    padding-top: 10px;
}
.PropertySearcherForm td:nth-child(odd) {
    text-align: right;
    padding-right: 5px;
}
.PropertySearcherForm td:nth-child(even) {
    text-align: left;
}
.PropertySearcherForm td:nth-child(2) {
    padding-right: 100px;
}
.PropertySearcherForm tr:nth-child(3) {
    width: 100%;
    text-align: right;
}
.PropertySearcherForm input {
    width: 200px;
}
.PropertySearcherForm select {
    width: 200px;
}
.PropertySearcherForm .submit_button {
    width: 100px;
    height: 50px;
    background-color: darkgreen;
    color: white;
}