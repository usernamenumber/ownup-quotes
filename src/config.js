// It's not good practice to store things like api keys in env vars,
// so this will allow us to transparently swap over to another 
// mechanism for retrieving the key in the future.
export default {
    ownup: {
        api_key: process.env.REACT_APP_OWNUP_API_KEY,
        api_url: "https://ss6b2ke2ca.execute-api.us-east-1.amazonaws.com/Prod/quotes",
    }
}