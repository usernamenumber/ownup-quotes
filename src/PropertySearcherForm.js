import Axios from 'axios';
import { useState } from 'react';
import config from './config.js';

const PropertySearcherForm = (props) => {
    const [formData, setFormData] = useState({
        loanSize:     null,
        propertyType: 'SingleFamily',
        creditScore:  null,
        occupancy:    'Primary'
    });
    
    const onChangeHandler = (event) => {
        const {name, value} = event.target;
        setFormData({...formData, [name]: value});
    }

    const onSubmitHandler = (event) => {
        event.preventDefault();
        const {api_url, api_key} = config.ownup;
        const payload = {
            headers: {
                authorization: `OU-AUTH ${api_key}`,
            },
            params: formData,
        };
        Axios.get(api_url, payload)
        .catch((error) => {
            console.log(error);
        })
        .then((response) => {
            console.log(response);
        })
        console.log(`Submission handler ran with ${payload}`);
    }

    return (
      <div className="PropertySearcherForm">
        <form onSubmit={onSubmitHandler}>
            <table>
                <tbody>
                <tr>
                    <td><label htmlFor="loanSize">Loan Size</label></td>
                    <td><input id="loanSize" name="loanSize" onChange = {onChangeHandler}></input></td>
                    <td><label htmlFor="propertyType">Property Type</label></td>
                    <td><select id="propertyType" name="propertyType" onChange = {onChangeHandler}>
                        <option value="SingleFamily">Single Family</option>
                        <option value="MultiFamily">Multi-Family</option>
                        <option value="Condo">Condo</option>
                        <option value="Townhouse">Townhouse</option>
                    </select></td>
                </tr>
                <tr>
                    <td><label htmlFor="creditScore">Credit Score</label></td>
                    <td><input id="creditScore" name="creditScore" onChange = {onChangeHandler}></input></td>
                    <td><label htmlFor="occupancy">Occupancy</label></td>
                    <td><select id="occupancy" name="occupancy" onChange = {onChangeHandler}>
                        <option value="Primary">Primary</option>
                        <option value="Secondary">Secondary</option>
                        <option value="Investment">Investment</option>
                    </select></td>
                </tr>
                <tr>
                    <td colSpan="4"><input className="submit_button" type="submit" value="Quote Rates"/></td>
                </tr>
                </tbody>
            </table>
        </form>
      </div>
    );
  }
  
  export default PropertySearcherForm;