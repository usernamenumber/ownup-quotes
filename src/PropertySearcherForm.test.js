import Axios from 'axios';
import React from 'react';
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import renderer from 'react-test-renderer';
import PropertySearcherForm from './PropertySearcherForm.js';
import config from './config.js';

jest.mock('axios');
Enzyme.configure({adapter: new Adapter() });

it('renders correctly (matches snapshot)', () => {
  const tree = renderer
    .create(<PropertySearcherForm/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('PropertySearcherForm API interactions', () => {

  const ownUpReturnData = {
    "rateQuotes": [
      {
        "lenderName": "TFB Federal Credit Union",
        "loanType": "7/1 ARM",
        "interestRate": 3.5,
        "closingCosts": 2000,
        "monthlyPayment": 449.0446878088235,
        "apr": 3.660036729314016
      }
    ]    
  };

  let wrapper;
  beforeEach(() => {
    wrapper = Enzyme.mount(Enzyme.mount(<PropertySearcherForm />).get(0))
    Axios.get.mockResolvedValue(ownUpReturnData);
  });

  afterEach(() => {
      jest.clearAllMocks();
  });

  it('submits the expected API query', () => {
    const form = wrapper.find('form').first();
    const fields = {
      loanSize: "100000",
      propertyType: "Townhouse",
      creditScore: "600",
      occupancy: "Investment",
    }

    for ( let fieldName in fields ) {
      const val = fields[fieldName];
      const field = form.find('#' + fieldName).first();
      field.instance().value = val;
      field.simulate('change');
    }
    form.simulate('submit');

    const {api_key, api_url} = config.ownup;
    expect(Axios.get).toHaveBeenCalledWith(
      api_url,
      {
        headers: {
          authorization: `OU-AUTH ${api_key}`,
        },
        params: fields,
      },
    );
  });

  // it('stores the API response in state', () => {
  //   Axios.get.returnValue(Promise.resolve({
  //     "rateQuotes": [
  //       {
  //         "lenderName": "TFB Federal Credit Union",
  //         "loanType": "7/1 ARM",
  //         "interestRate": 3.5,
  //         "closingCosts": 2000,
  //         "monthlyPayment": 449.0446878088235,
  //         "apr": 3.660036729314016
  //       },
  //     ]
  //   }));
  // });
});