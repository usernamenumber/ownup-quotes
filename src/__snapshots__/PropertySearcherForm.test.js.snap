// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly (matches snapshot) 1`] = `
<div
  className="PropertySearcherForm"
>
  <form
    onSubmit={[Function]}
  >
    <table>
      <tbody>
        <tr>
          <td>
            <label
              htmlFor="loanSize"
            >
              Loan Size
            </label>
          </td>
          <td>
            <input
              id="loanSize"
              name="loanSize"
              onChange={[Function]}
            />
          </td>
          <td>
            <label
              htmlFor="propertyType"
            >
              Property Type
            </label>
          </td>
          <td>
            <select
              id="propertyType"
              name="propertyType"
              onChange={[Function]}
            >
              <option
                value="SingleFamily"
              >
                Single Family
              </option>
              <option
                value="MultiFamily"
              >
                Multi-Family
              </option>
              <option
                value="Condo"
              >
                Condo
              </option>
              <option
                value="Townhouse"
              >
                Townhouse
              </option>
            </select>
          </td>
        </tr>
        <tr>
          <td>
            <label
              htmlFor="creditScore"
            >
              Credit Score
            </label>
          </td>
          <td>
            <input
              id="creditScore"
              name="creditScore"
              onChange={[Function]}
            />
          </td>
          <td>
            <label
              htmlFor="occupancy"
            >
              Occupancy
            </label>
          </td>
          <td>
            <select
              id="occupancy"
              name="occupancy"
              onChange={[Function]}
            >
              <option
                value="Primary"
              >
                Primary
              </option>
              <option
                value="Secondary"
              >
                Secondary
              </option>
              <option
                value="Investment"
              >
                Investment
              </option>
            </select>
          </td>
        </tr>
        <tr>
          <td
            colSpan="4"
          >
            <input
              className="submit_button"
              type="submit"
              value="Quote Rates"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </form>
</div>
`;
